File: 0490-the-maze/0490-the-maze.cpp
class Solution {
private:
      vector<int> moves;
      vector<vector<int> > maze;
      vector<int> destination;    
public:
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        this->maze = maze;
        vector<vector<bool> > visited(maze.size(), vector<bool>(maze[0].size()));
        this->destination = destination;
        moves.push_back(-1);
        moves.push_back(0);
        moves.push_back(1);
        moves.push_back(0);
        moves.push_back(-1);
        return hasPathRecur(start[0], start[1], visited);
    }

    bool hasPathRecur(int x, int y, vector<vector<bool> >& visited) {
      if(visited[x][y]) {
        return false;
      }

      if(x == destination[0] && y == destination[1]) {
        return true;
      }

      visited[x][y] = true;
      for(int m = 0; m < moves.size() - 1; m++) {
        int newX = x;
        int newY = y;
        while(newX >= 0 && newX < maze.size()
                 && newY >= 0 && newY < maze[newX].size()
                      && maze[newX][newY] == 0){
           newX += moves[m];
           newY += moves[m + 1];
         }

         // going to previous step
         if(hasPathRecur(newX - moves[m], newY - moves[m + 1], visited)) {
           return true;
         }
      }

      return false;
    }
};

File: 0490-the-maze/0490-the-maze.go
var Moves = []int{-1, 0, 1, 0, -1}

func hasPath(maze [][]int, start []int, destination []int) bool {
    var visited = make([][]bool, len(maze))
	for i := range visited {
		visited[i] = make([]bool, len(maze[0]))
	}

	return hasPathRecur(maze, visited, start, destination)
}

func hasPathRecur(maze [][]int, visited [][]bool, start []int, destination []int) bool {
	if visited[start[0]][start[1]] {
		return false
	}
	if start[0] == destination[0] && start[1] == destination[1] {
		return true
	}
	visited[start[0]][start[1]] = true
	for move := 0; move < len(Moves)-1; move++ {
		var x = start[0]
		var y = start[1]

		for x >= 0 && x < len(maze) && y >= 0 && y < len(maze[0]) && maze[x][y] == 0 {
			x += Moves[move]
			y += Moves[move+1]
		}

		if hasPathRecur(maze, visited, []int{x - Moves[move], y - Moves[move+1]}, destination) {
			return true
		}
	}

	return false
}
